File: brainfuck.py
import sys

BYTE_LIMIT = 256
MEMORY_SIZE = 4096

if len(sys.argv) < 2:
    print("The source file wasn't provided")
    exit(1)

with open(sys.argv[1], "r") as f:
    source = f.read()

size = len(source)

memory = [0] * MEMORY_SIZE
pointer = 0
index = 0
while index < size:
    if source[index] == '>':
        pointer = (pointer + 1) % MEMORY_SIZE

    elif source[index] == '<':
        pointer = (pointer - 1) % MEMORY_SIZE

    elif source[index] == '+':
        memory[pointer] = (memory[pointer] + 1) % BYTE_LIMIT

    elif source[index] == '-':
        memory[pointer] = (memory[pointer] - 1) % BYTE_LIMIT

    elif source[index] == '.':
        sys.stdout.write(chr(memory[pointer]))

    elif source[index] == ',':
        memory[pointer] = int(sys.stdin.read(1)) % BYTE_LIMIT

    elif source[index] == '[':
        if memory[pointer] == 0:
            scope = 1
            while scope != 0:
                index += 1
                if index == size:
                    break

                if source[index] == '[':
                    scope += 1

                if source[index] == ']':
                    scope -= 1

    elif source[index] == ']':
        if memory[pointer] != 0:
            scope = 1
            while scope != 0:
                index -= 1
                if index == 0:
                    break

                if source[index] == '[':
                    scope -= 1

                if source[index] == ']':
                    scope += 1

    index += 1

File: Brainfuck.cs
using System;
using System.IO;

internal class Program
{
    private const int MemorySize = 4096;

    private static int Main(string[] args)
    {
        if (args.Length == 0)
        {
            Console.WriteLine("The source file wasn't provided");
            return 1;
        }

        string source;
        try
        {
            source = File.ReadAllText(args[0]);
        }
        catch
        {
            Console.WriteLine("The source file couldn't read");
            return 1;
        }

        byte[] memory = new byte[MemorySize];
        int pointer = 0;
        int index, scope;
        for (index = 0; index < source.Length; index++)
        {
            switch (source[index])
            {
                case '>':
                    pointer++;
                    break;

                case '<':
                    pointer--;
                    break;
            
                case '+':
                    memory[pointer]++;
                    break;
            
                case '-':
                    memory[pointer]--;
                    break;

                case '.':
                    Console.Write((char)memory[pointer]);
                    break;
            
                case ',':
                    memory[pointer] = (byte)Console.Read();
                    break;
            
                case '[':
                    if (memory[pointer] == 0)
                    {
                        scope = 1;
                        while (scope != 0)
                        {
                            index++;
                            if (index == source.Length)
                                break;
                        
                            if (source[index] == '[')
                                scope++;
                        
                            if (source[index] == ']')
                                scope--;
                        }
                    }

                    break;

                case ']':
                    if (memory[pointer] != 0)
                    {
                        scope = 1;
                        while (scope != 0)
                        {
                            index--;
                            if (index == -1)
                                break;
                        
                            if (source[index] == '[')
                                scope--;
                        
                            if (source[index] == ']')
                                scope++;
                        }
                    }

                    break;
            }
        }

        return 0;
    }
}
